src/plugins/intel_gpu/src/plugin/multi_tensor_variable_state.cpp

```diff
@@ -0,0 +1,154 @@
+
+VariableStateIndirectKVCache::VariableStateIndirectKVCache(const VariableStateInfo& info,
+                                                           RemoteContextImpl::Ptr context,
+                                                           std::shared_ptr<cldnn::ShapePredictor> shape_predictor,
+                                                           size_t beam_idx,
+                                                           size_t concat_idx)
+    : MultiTensorState { {info}, context, shape_predictor}
+    , m_beam_idx(beam_idx)
+    , m_concat_idx(concat_idx) {
+    cldnn::layout beam_table_layout(get_beam_table_shape(info.m_layout.get_partial_shape()), ov::element::i32, cldnn::format::bfyx);
+    VariableStateInfo beam_table_state_info(info.m_id + "/beam_table", beam_table_layout);
+    m_hidden_states.push_back(std::make_shared<VariableState>(beam_table_state_info, context, shape_predictor));
+    OPENVINO_ASSERT(m_hidden_states.size() == 2, "[GPU] VariableStateIndirectKVCache expects 2 internal states to be initialized");
+}
+
+void VariableStateIndirectKVCache::reset() {
+    for (auto& state : m_hidden_states) {
+        state->reset();
+    }
+    m_is_set = false;
+}
+
+cldnn::memory::ptr VariableStateIndirectKVCache::get_memory() const {
+    return m_hidden_states[0]->get_memory();
+}
+
+const cldnn::layout& VariableStateIndirectKVCache::get_layout() const {
+    return m_hidden_states[0]->get_layout();
+}
+
+void VariableStateIndirectKVCache::set_state(const ov::SoPtr<ov::ITensor>& state) {
+    OPENVINO_ASSERT(m_hidden_states.size() == 2, "[GPU] Corrupted VariableStateIndirectKVCache. Expected 2 internal states. Got: ", m_hidden_states.size());
+    m_hidden_states[0]->set_state(state); // user can set only KV cache
+
+    // Beam table is reset to cleanup rearranges history
+    cldnn::layout bt_layout(get_beam_table_shape(state->get_shape()), ov::element::i32, cldnn::format::bfyx);
+    m_hidden_states[1]->reset();
+    m_hidden_states[1]->set_layout(bt_layout);
+}
+
+template<typename T>
+void copy_element(const void* src, void* dst, size_t src_offset, size_t dst_offset) {
+    static_cast<T*>(dst)[dst_offset] = static_cast<const T*>(src)[src_offset];
+}
+
+static void rearrange_cache(cldnn::memory::ptr kv_in_mem, cldnn::memory::ptr bt_mem, cldnn::memory::ptr kv_out_mem, cldnn::stream& stream) {
+    auto kv_layout = kv_in_mem->get_layout();
+    auto kv_shape = kv_layout.get_shape();
+    cldnn::mem_lock<uint8_t, cldnn::mem_lock_type::read> kv_in_ptr(kv_in_mem, stream);
+    cldnn::mem_lock<int32_t, cldnn::mem_lock_type::read> bt_in_ptr(bt_mem, stream);
+    cldnn::mem_lock<uint8_t, cldnn::mem_lock_type::write> kv_out_ptr(kv_out_mem, stream);
+
+    for (size_t b = 0; b < kv_shape[0]; b++) {
+        for (size_t f = 0; f < kv_shape[1]; f++) {
+            for (size_t y = 0; y < kv_shape[2]; y++) {
+                for (size_t x = 0; x < kv_shape[3]; x++) {
+                    size_t b_kv = bt_in_ptr[b* kv_shape[2] + y];
```

Comment:
Will concatenation always be along axis 2 in all scenarios? If not, it's better to add a check for this I think

Reply:
hardcoded index is replaced with parameter